src/support/err.c

```diff
@@ -504,12 +502,11 @@ __wt_illegal_value(WT_SESSION_IMPL *session, const char *name)
 	    name == NULL ? "" : name, name == NULL ? "" : ": ",
 	    "encountered an illegal file format or internal value");
 
-#if !defined(HAVE_DIAGNOSTIC)
-	return (__wt_panic(session));
+#if defined(HAVE_DIAGNOSTIC)
+	__wt_abort(session);                    /* Drop core if testing. */
 #endif
 
-	__wt_abort(session);			/* Drop core if testing. */
-	/* NOTREACHED */
+	return (WT_ERROR);
```

Comment:
@michaelcahill,

> __wt_illegal_value still returns an error, the only code that will run afterwards is error handling.

We have to call `__wt_panic` in order to shut the system down (set `WT_CONN_PANIC`), what prevents other code from running?

> In this case, a block read should fail, but that shouldn't bring down the system: it just means there is a file we can't open. That seems very different to me to a programming error where we control all of the possible values that could go into a switch.

Yes and no. If we can't open a file and read the first block, I agree we want to fail gracefully, it's likely a configuration/encryption problem. If we're running for a day and suddenly a block read fails, I think it's a pretty big problem and we want to shut down fast, before more damage occurs.

> My concern about the panic here (and more generally, the number of calls to panic through the code) is that I can't reason about whether an operation might cause a panic or not.

I know it's not helpful, but I think the rule is that any significant operation can panic: if a memory error ever corrupts `WT_PAGE.type`, we're almost certain to do a switch on the page type somewhere in every significant operation, fall into `WT_ILLEGAL_VALUE_ERR` in a switch statement, and panic. I can't think of any alternative, either, if there's a corrupted page in memory, we're done.

The way we handled this for encryption was to explicitly call verify on the root page when we opened each new file, and fail (relatively) gracefully. (It occurs to me we're probably still outputting a scary message into MongoDB's log, which may be a problem? I'll take a look in a minute.)

Is it the metadata file is special and we need to do something similar there?